cumulus/pallets/xcmp-queue/src/lib.rs

```diff
@@ -943,84 +575,182 @@ impl<T: Config> Pallet<T> {
 		iter: I,
 		max_weight: Weight,
 	) -> Weight {
-		let mut status = <InboundXcmpStatus<T>>::get();
-
-		let QueueConfigData { suspend_threshold, drop_threshold, .. } = <QueueConfig<T>>::get();
+		let mut meter = WeightMeter::with_limit(max_weight);
 
-		for (sender, sent_at, data) in iter {
-			// Figure out the message format.
-			let mut data_ref = data;
-			let format = match XcmpMessageFormat::decode_with_depth_limit(
-				MAX_XCM_DECODE_DEPTH,
-				&mut data_ref,
-			) {
+		for (sender, _sent_at, mut data) in iter {
+			let format = match XcmpMessageFormat::decode(&mut data) {
 				Ok(f) => f,
 				Err(_) => {
-					debug_assert!(false, "Unknown XCMP message format. Silently dropping message");
+					defensive!("Unknown XCMP message format - dropping");
 					continue
 				},
 			};
-			if format == XcmpMessageFormat::Signals {
-				while !data_ref.is_empty() {
-					use ChannelSignal::*;
-					match ChannelSignal::decode(&mut data_ref) {
-						Ok(Suspend) => Self::suspend_channel(sender),
-						Ok(Resume) => Self::resume_channel(sender),
-						Err(_) => break,
-					}
-				}
-			} else {
-				// Record the fact we received it.
-				match status.binary_search_by_key(&sender, |item| item.sender) {
-					Ok(i) => {
-						let count = status[i].message_metadata.len();
-						if count as u32 >= suspend_threshold && status[i].state == InboundState::Ok
+
+			match format {
+				XcmpMessageFormat::Signals =>
+					while !data.is_empty() {
+						if meter
+							.try_consume(
+								T::WeightInfo::suspend_channel()
+									.max(T::WeightInfo::resume_channel()),
+							)
+							.is_err()
 						{
-							status[i].state = InboundState::Suspended;
-							let r = Self::send_signal(sender, ChannelSignal::Suspend);
-							if r.is_err() {
-								log::warn!(
-									"Attempt to suspend channel failed. Messages may be dropped."
-								);
-							}
+							defensive!("Not enough weight to process signals - dropping");
+							break
 						}
-						if (count as u32) < drop_threshold {
-							status[i].message_metadata.push((sent_at, format));
-						} else {
-							debug_assert!(
-								false,
-								"XCMP channel queue full. Silently dropping message"
-							);
+
+						match ChannelSignal::decode(&mut data) {
+							Ok(ChannelSignal::Suspend) => Self::suspend_channel(sender),
+							Ok(ChannelSignal::Resume) => Self::resume_channel(sender),
+							Err(_) => {
+								defensive!("Undecodable channel signal - dropping");
+								break
+							},
 						}
 					},
-					Err(_) => status.push(InboundChannelDetails {
-						sender,
-						state: InboundState::Ok,
-						message_metadata: vec![(sent_at, format)],
-					}),
-				}
-				// Queue the payload for later execution.
-				<InboundXcmpMessages<T>>::insert(sender, sent_at, data_ref);
+				XcmpMessageFormat::ConcatenatedVersionedXcm => {
+					while !data.is_empty() {
+						let Ok(xcm) = Self::split_concatenated_xcm(&mut data, &mut meter) else {
```

Comment:
Yes it is preferable to not do this en/decode stuff here since it is wasteful.  
When we upgrade the HRMP channel protocol to the next version, then we can use double-encoded messages such that no en/decoding is needed in this split function.

PS: Renamed to `take_first_concatenated_xcm`.